src/display/annotation_layer.js

```diff
@@ -407,6 +439,93 @@ class TextAnnotationElement extends AnnotationElement {
   }
 }
 
+class FreeTextAnnotationElement extends AnnotationElement {
+  constructor(parameters) {
+    const isRenderable = !!(
+      parameters.data.hasPopup ||
+      parameters.data.title ||
+      parameters.data.contents
+    );
+    super(parameters, isRenderable, /* ignoreBorder = */ true);
+  }
+
+  /**
+   * Render the free text annotation's HTML element in the empty container.
+   *
+   * @public
+   * @memberof FreeTextAnnotationElement
+   * @returns {HTMLSectionElement}
+   */
+  render() {
+    this.container.className = "freeTextAnnotation";
+    const data = this.data;
+
+    if (!this.data.hasPopup) {
+      this._createPopup(this.container, null, this.data);
+    }
+
+    const div = document.createElement("div");
+    const divWidth = data.rect[2] - data.rect[0];
+    const divHeight = data.rect[3] - data.rect[1];
+    let style = "width: " + divWidth + "px;";
+    style += "height: " + divHeight + "px;";
+
+    if (data.hasAppearance) {
+      // The annotation is already rendering by stream appareance,
+      // we should only create popup if necessary
+      div.setAttribute("style", style);
+      this.container.append(div);
+
+      if (!this.data.hasPopup) {
+        this._createPopup(this.container, div, data);
```

Comment:
We should only create the popup once and it's already happening above. I think the question is also here: why does the separate `div` need to be created? Can't `this.container` just be used considering it's also already of the correct width/height so you don't have to calculate that yourself anymore?

Reply:
I also think the fact that multiple popups are rendered is what's breaking the reference tests.

Reply:
Once I commented out the first popup, the reftests were no longer broken. Thanks!
I do still see some minor differences between the reference and test reftests, but they seem very minor.

Reply:
> I do still see some minor differences between the reference and test reftests, but they seem very minor.

Even if changes in the reference tests are minor, please make sure that you can explain *why* they happen.
(Unexpected changes, without further investigation, *could* maybe point to something being wrong.)